### PR TITLE
Addressing spec inconsistencies with Open API repo spec (v2.3.0)

### PR DESCRIPTION
**Pull request by sergiishamrai-okta**
Java SDK uses a local copy of the Open API spec, by the original design. This PR contains fixes in swagger code generator classes and templates to address all inconsistencies with OpenAPI spec.

## Issue(s)
OKTA-349906

## Category

- [x] Bugfix
- [x] Enhancement
- [x] Unit Test(s)

### CODE REVIEW

File: swagger-templates/src/main/resources/OktaJava/modelInnerEnum.mustache [hunk 1]
```diff
@@ -13,36 +13,36 @@
...
-    {{/gson}}
-    {{^gson}}
-        {{#allowableValues}}
-            {{#enumVars}}
-    {{{name}}}({{{value}}}){{^-last}},
-    {{/-last}}{{#-last}};{{/-last}}
-            {{/enumVars}}
-        {{/allowableValues}}
-    {{/gson}}
+    /**
```
Comment (arvindkrishnakumar-okta) [inline swagger-templates/src/main/resources/OktaJava/modelInnerEnum.mustache:16]: No changes except for white space. Restore this file so it doesn't show up in the diff.

Comment (sergiishamrai-okta) [reply swagger-templates/src/main/resources/OktaJava/modelInnerEnum.mustache:16]: This file is not well-formatted. Code generated by this template shifted and looks not pretty.

**Comment by sergiishamrai-okta** [comment]
PRs to review
https://github.com/okta/okta-management-openapi-spec/pull/28
https://github.com/okta/okta-management-openapi-spec/pull/31
https://github.com/okta/okta-management-openapi-spec/pull/30

File: impl/src/main/java/com/okta/sdk/impl/resource/AbstractPropertyRetriever.java [hunk 2]
```diff
@@ -202,6 +202,10 @@ protected Date getDateProperty(DateProperty key) {
             return null;
         }
 
+        if(value instanceof Date) {
```
Comment (arvindkrishnakumar-okta) [inline impl/src/main/java/com/okta/sdk/impl/resource/AbstractPropertyRetriever.java:205]: Why is this needed now?

File: impl/src/main/java/com/okta/sdk/impl/resource/DefaultUserBuilder.java [hunk 3]
```diff
@@ -18,15 +18,8 @@
...
-import com.okta.sdk.resource.user.PasswordCredentialHook;
-import com.okta.sdk.resource.user.RecoveryQuestionCredential;
-import com.okta.sdk.resource.user.UserBuilder;
-import com.okta.sdk.resource.user.PasswordCredential;
-import com.okta.sdk.resource.user.User;
-import com.okta.sdk.resource.user.UserCredentials;
-import com.okta.sdk.resource.user.UserNextLogin;
-import com.okta.sdk.resource.user.UserProfile;
-import com.okta.sdk.resource.user.UserType;
+import com.okta.sdk.resource.user.*;
```
Comment (arvindkrishnakumar-okta) [inline impl/src/main/java/com/okta/sdk/impl/resource/DefaultUserBuilder.java:21]: Nit: Put back the single imports and avoid wildcards.

File: pom.xml [hunk 4]
```diff
@@ -343,8 +343,8 @@
                         </oldVersion>
                         <parameter>
                             <onlyModified>true</onlyModified>
-                            <breakBuildOnBinaryIncompatibleModifications>true</breakBuildOnBinaryIncompatibleModifications>
-                            <breakBuildBasedOnSemanticVersioning>true</breakBuildBasedOnSemanticVersioning>
+                            <breakBuildOnBinaryIncompatibleModifications>false</breakBuildOnBinaryIncompatibleModifications>
```
Comment (arvindkrishnakumar-okta) [inline pom.xml:346]: Create a followup JIRA item to revert this after the major release.

File: swagger-templates/src/main/java/com/okta/swagger/codegen/AbstractOktaJavaClientCodegen.java [hunk 5]
```diff
@@ -415,6 +423,11 @@ else if ("read".equals(alias) || "create".equals(alias)) {
                                 }
                             });
 
+                            //do not implement interface Deletable when delete method has some arguments
```
Comment (arvindkrishnakumar-okta) [inline swagger-templates/src/main/java/com/okta/swagger/codegen/AbstractOktaJavaClientCodegen.java:426]: Why so? Can you add some code comments explaining why?

**Comment by arvindkrishnakumar-okta** [comment]
Can you attach a diff of the API changes reported by japicmp plugin?

**Comment by arvindkrishnakumar-okta** [comment]
Please add a note to the PR description on which Open API spec version we're syncing the Java SDK spec to?

File: pom.xml (hunk 4 above)
Comment (sergiishamrai-okta) [reply pom.xml:346]: Internal Ref: [OKTA-379452](https://oktainc.atlassian.net/browse/OKTA-379452)

File: impl/src/main/java/com/okta/sdk/impl/resource/AbstractPropertyRetriever.java (hunk 2 above)
Comment (sergiishamrai-okta) [reply impl/src/main/java/com/okta/sdk/impl/resource/AbstractPropertyRetriever.java:205]: This method not working properly because of

`dateFormatter.parse(String.valueOf(value));`

If `value` is a string formatted as `"yyyy-MM-dd'T'HH:mm:ss.SSSZ"` then its works fine, but when the `value` is `Date` the result of `String.valueOf(value)` is a string formatted as '"EEE MMM dd HH:mm:ss zzz yyyy"'

So this method will crash in this case

File: swagger-templates/src/main/java/com/okta/swagger/codegen/AbstractOktaJavaClientCodegen.java (hunk 5 above)
Comment (sergiishamrai-okta) [reply swagger-templates/src/main/java/com/okta/swagger/codegen/AbstractOktaJavaClientCodegen.java:426]: `Deletable` defined as public interface with only one method `void delete();` and implementable only for objects with CRUD operation `delete()`

But for objects like `AuthorizationServerPolicy`  `delete` method requere 2 args: `authServerId` and `policyId`. In this case `policyId` will be taken from `src: id` but `authServerId` should be passed manually, so this object is not self-deletable and should not implement the `Deletable` interface

**Comment by sergiishamrai-okta** [comment]
Japicmp report: 

[japicmp.txt](https://github.com/okta/okta-sdk-java/files/6155488/japicmp.txt)

**Comment by arvindkrishnakumar-okta** [comment]
> Japicmp report:
> 
> [japicmp.txt](https://github.com/okta/okta-sdk-java/files/6155488/japicmp.txt)

html file please? it would be more readable.

File: MIGRATING.md [hunk 6]
```diff
@@ -2,6 +2,315 @@
...
+    - First argument type changed from `User` to `CreateUserRequest`
+
+Below methods have been added.
+- `UserSchema updateApplicationUserProfile(String appInstanceId, UserSchema userSchema)`
+- `UserSchema updateApplicationUserProfile(String appInstanceId)`
+- `UserSchema updateUserProfile(String schemaId, UserSchema userSchema)`
+    
+### Package `com.okta.sdk.resource.user.type.UserType`
+
+Class `com.okta.sdk.resource.user.UserType` moved under `com.okta.sdk.resource.user.type.UserType`
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:24]:
```suggestion
Class `com.okta.sdk.resource.user.UserType` moved to `com.okta.sdk.resource.user.type.UserType`
```

File: MIGRATING.md [hunk 7]
```diff
@@ -2,6 +2,315 @@
...
+
+### Package `com.okta.sdk.resource.user.type.UserTypeList`
+
+Class `com.okta.sdk.resource.user.UserTypeList` moved under `com.okta.sdk.resource.user.type.UserTypeList`
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:28]:
```suggestion
Class `com.okta.sdk.resource.user.UserTypeList` moved to `com.okta.sdk.resource.user.type.UserTypeList`
```

File: MIGRATING.md [hunk 8]
```diff
@@ -2,6 +2,315 @@
...
+
+### Package `com.okta.sdk.ds.RequestBuilder`
+
+Below method has been added.
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:32]: This method has already been introduced in `3.1.0`. So, this should not appear here. You should set the pom.xml version to `4.0.0-SNAPSHOT`, set the `okta.sdk.previousVersion` to `3.1.0` and run the japicmp report again.

File: MIGRATING.md [hunk 9]
```diff
@@ -2,6 +2,315 @@
...
+Below methods have been added.
+
+- `SingleLogout getSlo()`
+- `SpCertificate getSpCertificate()`
+- `SamlApplicationSettingsSignOn setSlo(SingleLogout slo)`
+- `SamlApplicationSettingsSignOn setSpCertificate(SpCertificate spCertificate)`
+
+### Package `com.okta.sdk.resource.application.SamlApplicationV1`
+
+The Interface has been removed.
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:80]: Name the interface that has been removed.

File: MIGRATING.md [hunk 10]
```diff
@@ -2,6 +2,315 @@
...
+### Package `com.okta.sdk.resource.application.SpCertificate`
+
+This is a newly created interface with methods listed below.
+
+- `List<String> getX5c()`
+- `SpCertificate setX5c(List<String> x5c)`
+
+### Package `com.okta.sdk.resource.authorization.server.AuthorizationServerPolicy`
+
+Below method has been renamed to keep naming consistency.
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:104]:
```suggestion
Below method has been renamed in the interest of naming consistency.
```

File: MIGRATING.md [hunk 11]
```diff
@@ -2,6 +2,315 @@
...
+- Renamed `deletePolicy` to `delete`
+
+Below method has undergone a signature change.
+- `void deletePolicyRule(String ruleId)` signature changed to `void deletePolicyRule(String authServerId, String ruleId)`
+
+Below method has undergone a signature change and renamed to keep naming consistency.
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:110]:
```suggestion
Below method has undergone a signature change in the interest of naming consistency.
```

File: MIGRATING.md [hunk 12]
```diff
@@ -2,6 +2,315 @@
...
+- `AuthorizationServerPolicy updatePolicy(String authServerId, AuthorizationServerPolicy authServerPolicy)` changed to `AuthorizationServerPolicy update(String authServerId)`
+
+### Package `com.okta.sdk.resource.authorization.server.policy.AuthorizationServerPolicyRule`
+
+Below method has been renamed to keep naming consistency.
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:115]: Below method has been renamed in the interest of naming consistency.

File: MIGRATING.md [hunk 13]
```diff
@@ -2,6 +2,315 @@
...
+### Package `com.okta.sdk.resource.inline.hook.InlineHookChannelConfig`
+
+Below methods have been added.
+
+- `String getMethod()`
+- `InlineHookChannelConfig setMethod(String method)`
+
+### Package `com.okta.sdk.resource.linked.object.LinkedObject`
+
+Below methods have been removed.
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:127]: Please add a note explaining why and what the callers already using this should do now?

File: MIGRATING.md [hunk 14]
```diff
@@ -2,6 +2,315 @@
...
+- `String getName()`
+- `LinkedObject setName(String name)`
+
+### Package `com.okta.sdk.resource.log.LogCredentialProvider`
+
+Enum `LogCredentialProvider` has the below field defined:
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:133]:
```suggestion
Enum `LogCredentialProvider` has the below new field definition:
```

File: MIGRATING.md [hunk 15]
```diff
@@ -2,6 +2,315 @@
...
+- `OKTA_CREDENTIAL_PROVIDER("OKTA_CREDENTIAL_PROVIDER")`
+
+### Package `com.okta.sdk.resource.policy.DevicePolicyRuleConditionPlatform$SupportedMDMFrameworksEnum`
+
+Interface `com.okta.sdk.resource.policy.MDMFrameworks` moved under `com.okta.sdk.resource.policy.DevicePolicyRuleConditionPlatform$SupportedMDMFrameworksEnum`
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:138]:
```suggestion
Interface `com.okta.sdk.resource.policy.MDMFrameworks` moved to `com.okta.sdk.resource.policy.DevicePolicyRuleConditionPlatform$SupportedMDMFrameworksEnum`
```

File: MIGRATING.md [hunk 16]
```diff
@@ -2,6 +2,315 @@
...
+
+### Package `com.okta.sdk.resource.policy.DevicePolicyRuleConditionPlatform$TypesEnum`
+
+Interface `com.okta.sdk.resource.policy.Platforms` moved under `com.okta.sdk.resource.policy.DevicePolicyRuleConditionPlatform$TypesEnum`
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:142]:
```suggestion
Interface `com.okta.sdk.resource.policy.Platforms` moved to `com.okta.sdk.resource.policy.DevicePolicyRuleConditionPlatform$TypesEnum`
```

File: MIGRATING.md [hunk 17]
```diff
@@ -2,6 +2,315 @@
...
+
+### Package `com.okta.sdk.resource.user.factor.PushUserFactor` 
+
+Below method has been added
+- `PushUserFactor setExpiresAt(Date expiresAt)`
+
+### Package `com.okta.sdk.resource.user.factor.UserFactor`
+
+Below method has been renamed to keep naming consistency.
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:151]:
```suggestion
Below method has been renamed in the interest of naming consistency.
```

File: MIGRATING.md [hunk 18]
```diff
@@ -2,6 +2,315 @@
...
+- `UserSchemaPublic setProperties(Map<String, Object> properties)`
+- `UserSchemaPublic setRequired(List<String> required)`
+- `UserSchemaPublic setType(String type)`
+
+### Package `com.okta.sdk.resource.user.User`
+
+Below method has undergone a signature change and renamed to keep naming consistency.
+- `void addAllAppsAsTargetToRole()` to `void addAllAppsAsTarget(String roleId)`
+- `void deleteFactor()` to `void deleteFactor(String factorId)`
+
```
Comment (arvindkrishnakumar-okta) [inline MIGRATING.md:309]: Remove empty lines.

**Review by arvindkrishnakumar-okta — CHANGES_REQUESTED** [review]
We should also write ITs for the new APIs that are brought in with this spec update (wherever possible).

File: pom.xml (hunk 4 above)
Comment (arvindkrishnakumar-okta) [inline pom.xml:346]: Now that you've updated the PR with next major (`4.0.0`), you should revert these back to `true` and the build should be fine.

**Review by arvindkrishnakumar-okta — APPROVED** [review]
LGTM, please double check if we've added ITs (wherever possible) for all the new models introduced with this spec update.
